File: src/ts/Models/Stores/GenreStore.ts
import * as _ from 'lodash';
import Libraries from '../../Libraries';
import StoreBase from './StoreBase';
import ApiMethods from '../../Definitions/ApiMethods';
import MopidyRef from '../Entities/MopidyRef';
import Genre from '../Entities/Genre';

export default class GenreStore extends StoreBase<Genre> {

    public async Init(): Promise<boolean> {
        const entities: Genre[] = [];

        const result = await this.Query(ApiMethods.LibraryBrowse, {
            uri: 'local:directory?type=genre'
        });
        const refs: MopidyRef[] = result.result;

        _.each(refs, async (ref) => {
            const genre = new Genre(ref.name, ref.uri);
            entities.push(genre);

            //const related = await this.Query(ApiMethods.LibraryBrowse, {
            //    uri: 'local:directory?type=artist'
            //});
        });

        this.Entities = Libraries.Enumerable.from(entities);

        return true;
    }
}

File: aspCore/Models/Genres/GenreStore.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using MusicFront.Models.Bases;
using MusicFront.Models.Mopidies.Methods.Libraries;
using System.Collections.Generic;
using System.Linq;

namespace MusicFront.Models.Genres
{
    public class GenreStore : StoreBase<Genre>
    {
        private const string QueryString = "local:directory?type=genre";

        public GenreStore([FromServices] Dbc dbc) : base(dbc)
        {
        }

        public Genre Get(int genreId)
            => this.Dbc.GetGenreQuery().FirstOrDefault(e => e.Id == genreId);

        public List<Genre> FindAll(string[] names, int[] ids)
        {
            var query = this.Dbc.GetGenreQuery();
            if (names != null && 0 < names.Length)
                query = query.Where(e => names.All(name => e.LowerName.Contains(name.ToLower())));
            if (ids != null && 0 < ids.Length)
                query = query.Where(e => ids.Contains(e.Id));

            return query.ToList();
        }

        public List<Artists.Artist> GetArtistsByGenre(Genre genre)
            =>  this.Dbc.GetArtistQuery()
                .Where(e => e.GenreArtists.Select(e2 => e2.GenreId).Contains(genre.Id))
                .OrderBy(e => e.Name)
                .ToList();

        public List<Albums.Album> GetAlbumsByGenre(Genre genre)
            =>  this.Dbc.GetAlbumQuery()
                .Where(e => e.GenreAlbums.Select(e2 => e2.GenreId).Contains(genre.Id))
                .OrderBy(e => e.Name)
                .ToList();

        public void Refresh()
        {
            this.Dbc.Genres.RemoveRange(this.Dbc.Genres);
            this.Dbc.SaveChanges();

            var result = Browse.Request(GenreStore.QueryString)
                .GetAwaiter()
                .GetResult();

            var genres = result.Select(e => new Genre()
            {
                Name = e.Name,
                LowerName = e.Name.ToLower(),
                Uri = e.Uri
            }).ToArray();

            this.Dbc.Genres.AddRange(genres);
            this.Dbc.SaveChanges();
        }
    }
}
